Error in make_spherical_differential function

When `make_spherical_differential` function is called with template parameters to return other quantity and `spherical differential` as argument then it considers it as a `cartesian` object and gives error message `static assertion failed: All components must have same dimensions
` .

> can you please provide an example and error message

Example:
```
quantity<degree_plane_angle> a(10.0* degrees);
quantity<plane_angle> b(1.5* radians);
quantity<velocity> c(70.0* meters/seconds);
auto motion1 = make_spherical_differential(a,b,c);
auto motion2 = make_spherical_differential
        <double,quantity<plane_angle>,quantity<plane_angle>,quantity<velocity>>(motion1);
```

Error:
```
astronomy/include/boost/astronomy/coordinate/cartesian_differential.hpp: In instantiation of ‘struct boost::astronomy::coordinate::cartesian_differential<double, boost::units::quantity<boost::units::unit<boost::units::list<boost::units::dim<boost::units::plane_angle_base_dimension, boost::units::static_rational<1> >, boost::units::dimensionless_type>, boost::units::homogeneous_system<boost::units::list<boost::units::si::meter_base_unit, boost::units::list<boost::units::scaled_base_unit<boost::units::cgs::gram_base_unit, boost::units::scale<10, boost::units::static_rational<3> > >, boost::units::list<boost::units::si::second_base_unit, boost::units::list<boost::units::si::ampere_base_unit, boost::units::list<boost::units::si::kelvin_base_unit, boost::units::list<boost::units::si::mole_base_unit, boost::units::list<boost::units::si::candela_base_unit, boost::units::list<boost::units::angle::radian_base_unit, boost::units::list<boost::units::angle::steradian_base_unit, boost::units::dimensionless_type> > > > > > > > > > > >, boost::units::quantity<boost::units::unit<boost::units::list<boost::units::dim<boost::units::plane_angle_base_dimension, boost::units::static_rational<1> >, boost::units::dimensionless_type>, boost::units::homogeneous_system<boost::units::list<boost::units::si::meter_base_unit, boost::units::list<boost::units::scaled_base_unit<boost::units::cgs::gram_base_unit, boost::units::scale<10, boost::units::static_rational<3> > >, boost::units::list<boost::units::si::second_base_unit, boost::units::list<boost::units::si::ampere_base_unit, boost::units::list<boost::units::si::kelvin_base_unit, boost::units::list<boost::units::si::mole_base_unit, boost::units::list<boost::units::si::candela_base_unit, boost::units::list<boost::units::angle::radian_base_unit, boost::units::list<boost::units::angle::steradian_base_unit, boost::units::dimensionless_type> > > > > > > > > > > >, boost::units::quantity<boost::units::unit<boost::units::list<boost::units::dim<boost::units::length_base_dimension, boost::units::static_rational<1> >, boost::units::list<boost::units::dim<boost::units::time_base_dimension, boost::units::static_rational<-1> >, boost::units::dimensionless_type> >, boost::units::homogeneous_system<boost::units::list<boost::units::si::meter_base_unit, boost::units::list<boost::units::scaled_base_unit<boost::units::cgs::gram_base_unit, boost::units::scale<10, boost::units::static_rational<3> > >, boost::units::list<boost::units::si::second_base_unit, boost::units::list<boost::units::si::ampere_base_unit, boost::units::list<boost::units::si::kelvin_base_unit, boost::units::list<boost::units::si::mole_base_unit, boost::units::list<boost::units::si::candela_base_unit, boost::units::list<boost::units::angle::radian_base_unit, boost::units::list<boost::units::angle::steradian_base_unit, boost::units::dimensionless_type> > > > > > > > > > > > >’:
ex11.cpp:81:88:   recursively required by substitution of ‘template<class OtherCoordinateType, class XQuantity, class YQuantity, class ZQuantity> boost::astronomy::coordinate::spherical_differential<OtherCoordinateType, boost::units::quantity<boost::units::unit<boost::units::list<boost::units::dim<boost::units::plane_angle_base_dimension, boost::units::static_rational<1> >, boost::units::dimensionless_type>, boost::units::homogeneous_system<boost::units::list<boost::units::si::meter_base_unit, boost::units::list<boost::units::scaled_base_unit<boost::units::cgs::gram_base_unit, boost::units::scale<10, boost::units::static_rational<3> > >, boost::units::list<boost::units::si::second_base_unit, boost::units::list<boost::units::si::ampere_base_unit, boost::units::list<boost::units::si::kelvin_base_unit, boost::units::list<boost::units::si::mole_base_unit, boost::units::list<boost::units::si::candela_base_unit, boost::units::list<boost::units::angle::radian_base_unit, boost::units::list<boost::units::angle::steradian_base_unit, boost::units::dimensionless_type> > > > > > > > > > >, CoordinateType>, boost::units::quantity<boost::units::unit<boost::units::list<boost::units::dim<boost::units::plane_angle_base_dimension, boost::units::static_rational<1> >, boost::units::dimensionless_type>, boost::units::homogeneous_system<boost::units::list<boost::units::si::meter_base_unit, boost::units::list<boost::units::scaled_base_unit<boost::units::cgs::gram_base_unit, boost::units::scale<10, boost::units::static_rational<3> > >, boost::units::list<boost::units::si::second_base_unit, boost::units::list<boost::units::si::ampere_base_unit, boost::units::list<boost::units::si::kelvin_base_unit, boost::units::list<boost::units::si::mole_base_unit, boost::units::list<boost::units::si::candela_base_unit, boost::units::list<boost::units::angle::radian_base_unit, boost::units::list<boost::units::angle::steradian_base_unit, boost::units::dimensionless_type> > > > > > > > > > >, CoordinateType>, XQuantity> boost::astronomy::coordinate::make_spherical_differential(const boost::astronomy::coordinate::cartesian_differential<ReturnCoordinateType, ReturnXQuantity, ReturnYQuantity, ReturnZQuantity>&) [with OtherCoordinateType = double; XQuantity = boost::units::quantity<boost::units::unit<boost::units::list<boost::units::dim<boost::units::plane_angle_base_dimension, boost::units::static_rational<1> >, boost::units::dimensionless_type>, boost::units::homogeneous_system<boost::units::list<boost::units::si::meter_base_unit, boost::units::list<boost::units::scaled_base_unit<boost::units::cgs::gram_base_unit, boost::units::scale<10, boost::units::static_rational<3> > >, boost::units::list<boost::units::si::second_base_unit, boost::units::list<boost::units::si::ampere_base_unit, boost::units::list<boost::units::si::kelvin_base_unit, boost::units::list<boost::units::si::mole_base_unit, boost::units::list<boost::units::si::candela_base_unit, boost::units::list<boost::units::angle::radian_base_unit, boost::units::list<boost::units::angle::steradian_base_unit, boost::units::dimensionless_type> > > > > > > > > > > >; YQuantity = boost::units::quantity<boost::units::unit<boost::units::list<boost::units::dim<boost::units::plane_angle_base_dimension, boost::units::static_rational<1> >, boost::units::dimensionless_type>, boost::units::homogeneous_system<boost::units::list<boost::units::si::meter_base_unit, boost::units::list<boost::units::scaled_base_unit<boost::units::cgs::gram_base_unit, boost::units::scale<10, boost::units::static_rational<3> > >, boost::units::list<boost::units::si::second_base_unit, boost::units::list<boost::units::si::ampere_base_unit, boost::units::list<boost::units::si::kelvin_base_unit, boost::units::list<boost::units::si::mole_base_unit, boost::units::list<boost::units::si::candela_base_unit, boost::units::list<boost::units::angle::radian_base_unit, boost::units::list<boost::units::angle::steradian_base_unit, boost::units::dimensionless_type> > > > > > > > > > > >; ZQuantity = boost::units::quantity<boost::units::unit<boost::units::list<boost::units::dim<boost::units::length_base_dimension, boost::units::static_rational<1> >, boost::units::list<boost::units::dim<boost::units::time_base_dimension, boost::units::static_rational<-1> >, boost::units::dimensionless_type> >, boost::units::homogeneous_system<boost::units::list<boost::units::si::meter_base_unit, boost::units::list<boost::units::scaled_base_unit<boost::units::cgs::gram_base_unit, boost::units::scale<10, boost::units::static_rational<3> > >, boost::units::list<boost::units::si::second_base_unit, boost::units::list<boost::units::si::ampere_base_unit, boost::units::list<boost::units::si::kelvin_base_unit, boost::units::list<boost::units::si::mole_base_unit, boost::units::list<boost::units::si::candela_base_unit, boost::units::list<boost::units::angle::radian_base_unit, boost::units::list<boost::units::angle::steradian_base_unit, boost::units::dimensionless_type> > > > > > > > > > > >]’
ex11.cpp:81:88:   required from here
astronomy/include/boost/astronomy/coordinate/cartesian_differential.hpp:36:5: error: static assertion failed: All components must have same dimensions
     BOOST_STATIC_ASSERT_MSG(
```

--- a/include/boost/astronomy/coordinate/spherical_differential.hpp
+++ b/include/boost/astronomy/coordinate/spherical_differential.hpp
@@ -395,75 +395,44 @@ make_spherical_differential
         >(pointObject);
 }
 
-//!constructs object from any type of cartesian differential
+//!constructs object from any type of differential
 template 
 <
-    typename OtherCoordinateType,
-    typename XQuantity,
-    typename YQuantity,
-    typename ZQuantity
+    typename OtherDifferential
 >
-spherical_differential
-<
-    OtherCoordinateType,
-    bu::quantity<bu::si::plane_angle, OtherCoordinateType>,
-    bu::quantity<bu::si::plane_angle, OtherCoordinateType>,
-    XQuantity
->
-make_spherical_differential
+auto make_spherical_differential
 (
-    cartesian_differential
-    <
-        OtherCoordinateType,
-        XQuantity,
-        YQuantity,
-        ZQuantity
-    > const& other
+    OtherDifferential const& other
 )
-{
+{   
+    auto temp = make_cartesian_differential(other);
+    typedef decltype(temp) cartesian_type;
+
     bg::model::point
         <
-            OtherCoordinateType,
+            typename cartesian_type::type,
             3,
             bg::cs::cartesian
         > tempPoint;
 
-    bg::set<0>(tempPoint, other.get_dx().value());
-    bg::set<1>(tempPoint, static_cast<XQuantity>(other.get_dy()).value());
-    bg::set<2>(tempPoint, static_cast<XQuantity>(other.get_dz()).value());
+    bg::set<0>(tempPoint, temp.get_dx().value());
+    bg::set<1>(tempPoint, static_cast<typename cartesian_type::quantity1>
+        (temp.get_dy()).value());
+    bg::set<2>(tempPoint, static_cast<typename cartesian_type::quantity1>
+        (temp.get_dz()).value());
 
-    bg::model::point<OtherCoordinateType, 3, bg::cs::spherical<radian>> result;
+    bg::model::point<typename cartesian_type::type, 3, bg::cs::spherical<radian>> result;
     bg::transform(tempPoint, result);
 
     return spherical_differential
         <
-            OtherCoordinateType,
-            bu::quantity<bu::si::plane_angle, OtherCoordinateType>,
-            bu::quantity<bu::si::plane_angle, OtherCoordinateType>,
-            XQuantity
+            typename cartesian_type::type,
+            bu::quantity<bu::si::plane_angle, typename cartesian_type::type>,
+            bu::quantity<bu::si::plane_angle, typename cartesian_type::type>,
+            typename cartesian_type::quantity1
         >(result);
 }
 
-//!constructs object from any type of spherical equatorial differential
-template <typename OtherDifferential>
-spherical_differential
-<
-    typename OtherDifferential::type,
-    typename OtherDifferential::quantity1,
-    typename OtherDifferential::quantity2,
-    typename OtherDifferential::quantity3
->
-make_spherical_differential(OtherDifferential const& other)
-{
-
-    return spherical_differential
-        <
-            typename OtherDifferential::type,
-            typename OtherDifferential::quantity1,
-            typename OtherDifferential::quantity2,
-            typename OtherDifferential::quantity3
-        >(other.get_differential());
-}
 }}}//namespace boost::astronomy::coordinate
 
 #endif // !BOOST_ASTRONOMY_COORDINATE_SPHERICAL_DIFFERENTIAL_HPP

--- a/include/boost/astronomy/coordinate/spherical_representation.hpp
+++ b/include/boost/astronomy/coordinate/spherical_representation.hpp
@@ -228,7 +228,7 @@ public:
             LatQuantity,
             LonQuantity,
             DistQuantity
-        > result = make_spherical_differential(temp);
+        > result = make_spherical_representation(temp);
 
         return result;
     }
@@ -371,43 +371,31 @@ make_spherical_representation
 //!Create cartesian_representation from other type of representations
 template
 <
-    typename CoordinateType,
-    typename XQuantity,
-    typename YQuantity,
-    typename ZQuantity
+    typename OtherRepresentation
 >
-spherical_representation
-<
-    CoordinateType,
-    bu::quantity<bu::si::plane_angle, CoordinateType>,
-    bu::quantity<bu::si::plane_angle, CoordinateType>,
-    XQuantity
-> make_spherical_representation
-(cartesian_representation<CoordinateType, XQuantity, YQuantity, ZQuantity> const& other)
+auto make_spherical_representation
+(
+    OtherRepresentation const& other
+)
 {
-    typedef cartesian_representation
-        <
-        CoordinateType,
-        XQuantity,
-        YQuantity,
-        ZQuantity
-        > cartesian_type;
+    auto temp = make_cartesian_representation(other);
+    typedef decltype(temp) cartesian_type;
 
     bg::model::point<typename cartesian_type::type, 3, bg::cs::cartesian> tempPoint;
     bg::model::point<typename cartesian_type::type, 3, bg::cs::spherical<radian>> result;
 
-    bg::set<0>(tempPoint, other.get_x().value());
+    bg::set<0>(tempPoint, temp.get_x().value());
     bg::set<1>
     (
         tempPoint,
         static_cast<typename cartesian_type::quantity1>
-        (other.get_y()).value()
+        (temp.get_y()).value()
     );
     bg::set<2>
     (
         tempPoint,
         static_cast<typename cartesian_type::quantity1>
-        (other.get_z()).value()
+        (temp.get_z()).value()
     );
 
     bg::transform(tempPoint, result);
@@ -419,28 +407,6 @@ spherical_representation
             bu::quantity<bu::si::plane_angle, typename cartesian_type::type>,
             typename cartesian_type::quantity1
         >(result);
-}
-
-//!Create spherical_equatorial_representation from other type of representations
-//A general implementation is used to avoid including spherical_equatorial_representation.hpp
-//error: using spherical_equatorial_representation before it's declaration
-template<typename OtherRepresentation>
-spherical_representation
-<
-    typename OtherRepresentation::type,
-    typename OtherRepresentation::quantity1,
-    typename OtherRepresentation::quantity2,
-    typename OtherRepresentation::quantity3
->
-make_spherical_representation(OtherRepresentation const& other)
-{
-    return spherical_representation
-        <
-        typename OtherRepresentation::type,
-        typename OtherRepresentation::quantity1,
-        typename OtherRepresentation::quantity2,
-        typename OtherRepresentation::quantity3
-        >(other.get_point());
 }
 
 }}} //namespace boost::astronomy::coordinate

